src/backend/cdb/cdbpath.c

```diff
@@ -1237,25 +1279,96 @@ cdbpath_motion_for_join(PlannerInfo *root,
 			}
 		}
 	}
+
 	/*
-	 * Replicated paths shouldn't occur except UPDATE/DELETE on replicated table.
+	 * Replicated paths shouldn't occur except UPDATE/DELETE on replicated
+	 * table or modifying CTEs with DML operations on replicated table with
+	 * SELECT command above. In first case when we UPDATE/DELETE replicated
+	 * table, the flag upd_del_replicated_table is set true, and, therefore,
+	 * we need to make any other probable locus replicated in order to achieve
+	 * correct update/delete. Another case - DML over replicated table inside
+	 * CTE with SELECT above.
 	 */
-	else if (CdbPathLocus_IsReplicated(outer.locus))
+	else if (CdbPathLocus_IsReplicated(outer.locus) ||
+			 CdbPathLocus_IsReplicated(inner.locus))
 	{
-		if (root->upd_del_replicated_table > 0)
-			CdbPathLocus_MakeReplicated(&inner.move_to,
-										CdbPathLocus_NumSegments(outer.locus));
+		CdbpathMfjRel *replicated;
+		CdbpathMfjRel *other;
+
+		if (CdbPathLocus_IsReplicated(outer.locus))
+		{
+			replicated = &outer;
+			other = &inner;
+		}
 		else
 		{
-			Assert(false);
-			goto fail;
+			replicated = &inner;
+			other = &outer;
 		}
-	}
-	else if (CdbPathLocus_IsReplicated(inner.locus))
-	{
+
+		/* Common UPDATE/DELETE on replicated table */
 		if (root->upd_del_replicated_table > 0)
-			CdbPathLocus_MakeReplicated(&outer.move_to,
-										CdbPathLocus_NumSegments(inner.locus));
+			CdbPathLocus_MakeReplicated(&other->move_to,
+								CdbPathLocus_NumSegments(replicated->locus));
+		else if (hasModifyingCTESelectRecur(root))
+		{
+			/* Entry has to be broadcasted */
+			if (CdbPathLocus_IsEntry(other->locus))
+				CdbPathLocus_MakeReplicated(&other->move_to,
+								CdbPathLocus_NumSegments(replicated->locus));
+			else if (CdbPathLocus_IsSingleQE(other->locus))
+			{
+				/*
+				 * bring locus Replicated to singleQE in order to perform join
+				 * on single segment
+				 */
+				int		numsegments = CdbPathLocus_CommonSegments(replicated->locus,
```

Comment:
Why not CdbPathLocus_MakeSingleQE to be easy?

Reply:
I blindly stole logic from SegmentGeneral [case]([same](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbpath.c#L1163-L1172)). There was an explicit locus assignment because the case included both Entry and SingleQE cases. And in my patch I separated these cases (Now, I consider it to be wrong).